lib/connection/connection.js

```diff
@@ -528,97 +528,142 @@ function merge(options1, options2) {
+  return;
 }
 
-function makeUnsecureConnection(self, family) {
-  // Create new connection instance
-  let connection_options;
-  if (self.domainSocket) {
-    connection_options = { path: self.host };
-  } else {
-    connection_options = { port: self.port, host: self.host };
-    connection_options.family = family;
-  }
-
-  const connection = net.createConnection(connection_options);
-
-  // Set the options for the connection
-  connection.setKeepAlive(self.keepAlive, self.keepAliveInitialDelay);
-  connection.setTimeout(self.connectionTimeout);
-  connection.setNoDelay(self.noDelay);
-
-  connection.once('connect', function() {
-    // Set socket timeout instead of connection timeout
-    connection.setTimeout(self.socketTimeout);
-    // Emit connect event
-    self.emit('connect', self);
-  });
+function fastFallback(self, _options, callback) {
+  const options = prepareConnectionOptions(self, _options);
+
+  let connection;
+  const connectionHandler = (err, _connection) => {
+    const _errorHandler = errorHandler(self);
+    const _timeoutHandler = timeoutHandler(self);
+    const _closeHandler = closeHandler(self);
+    const _dataHandler = dataHandler(self);
+
+    if (err) {
+      switch (err.type) {
+        case 'error':
+          if (err.family === 6) {
+            return function() {
+              if (self.logger.isDebug()) {
+                self.logger.debug(
+                  f(
+                    'connection %s for [%s:%s] errored out with [%s]',
+                    self.id,
+                    self.host,
+                    self.port,
+                    JSON.stringify(err)
+                  )
+                );
+              }
+            };
+          } else {
+            return _errorHandler(err.err);
+          }
+        case 'timeout':
+          return _timeoutHandler(err.err);
+        case 'close':
+          return _closeHandler(err.err);
+        case 'data':
+          return _dataHandler(err.err);
+      }
+    }
 
-  return connection;
-}
+    if (connection) {
+      _connection.removeAllListeners('error');
+      _connection.removeAllListeners('timeout');
+      _connection.removeAllListeners('close');
+      _connection.removeAllListeners('data');
+      _connection.end();
+      _connection.unref();
+    } else {
+      connection = _connection;
+    }
 
-function doConnect(self, family, _options, _errorHandler) {
-  self.connection = self.ssl
-    ? makeSSLConnection(self, _options)
-    : makeUnsecureConnection(self, family);
+    return callback(null, connection);
+  };
 
-  // Add handlers for events
-  self.connection.once('error', _errorHandler);
-  self.connection.once('timeout', timeoutHandler(self));
-  self.connection.once('close', closeHandler(self));
-  self.connection.on('data', dataHandler(self));
+  makeConnection(self, Object.assign({}, { family: 6 }, options), connectionHandler);
+  setTimeout(() => {
```

Comment:
why is this `setTimeout` here?

Reply:
This is part of the fast fallback algorithm, by having ipv4 wait 250ms to try to connect, it gives ipv6 preference.

Reply:
Ah okay! Can you leave a note indicating as much here for future readers?